### PR TITLE
Allocations: Unroll AdditionalFile extension method

### LINKED ISSUE #7440: Avoid calling Path.GetFileName in AdditionalFile
Accounts for 5037MB allocations in https://developercommunity.visualstudio.com/t/Visual-Studio-2022:-High-CPU-usage-by-Se/10378039

https://github.com/SonarSource/sonar-dotnet/blob/ab9c515a82f421588a480496f5138f5cfb3526b8/analyzers/src/SonarAnalyzer.Common/Extensions/AnalyzerOptionsExtensions.cs#L37

### PR DESCRIPTION
**Pull request by martin-strecker-sonarsource**
Fixes #7440

Tested with Akka.Net in a `dotnet build` context in a SQ EE SonaryWay run.

This change safes
* 5GB of string allocations caused by calls to `Path.GetFileName`
* 3GB of delegate allocations for the FirstOrDefault lambda argument
* 1,1 GB for the capture class for the filename parameter.

The total memory allocations are down to 32,5 GB from 41,7GB = 9,2 GB savings (22%)

The number of issues reported stayed the same (3514 warnings).

Before
image

After
image

### CODE REVIEW

File: analyzers/src/SonarAnalyzer.Common/Extensions/AnalyzerOptionsExtensions.cs [hunk 1]
```diff
@@ -33,6 +33,26 @@ public static class AnalyzerOptionsExtensions
     public static AdditionalText ProjectOutFolderPath(this AnalyzerOptions options) =>
         options.AdditionalFile("ProjectOutFolderPath.txt");
 
-    private static AdditionalText AdditionalFile(this AnalyzerOptions options, string fileName) =>
-        options.AdditionalFiles.FirstOrDefault(x => x.Path is not null && Path.GetFileName(x.Path).Equals(fileName, StringComparison.OrdinalIgnoreCase));
+    private static AdditionalText AdditionalFile(this AnalyzerOptions options, string fileName)
+    {
+        // HotPath: This code path needs tp be allocation free. Don't use Linq.
```
Comment (cristian-ambrosini-sonarsource) [inline analyzers/src/SonarAnalyzer.Common/Extensions/AnalyzerOptionsExtensions.cs:38]:
```suggestion
        // HotPath: This code path needs to be allocation free. Don't use Linq.
```

File: analyzers/src/SonarAnalyzer.Common/Extensions/AnalyzerOptionsExtensions.cs [hunk 2]
```diff
@@ -33,6 +33,26 @@ public static class AnalyzerOptionsExtensions
...
+        foreach (var additionalText in options.AdditionalFiles) // Uses the struct enumerator of ImmutableArray
+        {
+            // Don't use Path.GetFilename. It allocates a string.
+            if (additionalText.Path is { } path
+                && path.EndsWith(fileName, StringComparison.OrdinalIgnoreCase))
+            {
+                // The character before the filename (if there is a character) must be a directory separator
+                if (path.Length - fileName.Length - 1 is >= 0 and var separatorPosition
```
Comment (martin-strecker-sonarsource) [inline analyzers/src/SonarAnalyzer.Common/Extensions/AnalyzerOptionsExtensions.cs:46]: There is a test case for this: The additional file ends with `SonarLint.xml` but is named `FooBarSonarLint.xml`.
https://github.com/SonarSource/sonar-dotnet/blob/be9ca30695ba555f5d4b8c82c4b349bd53b37ddf/analyzers/tests/SonarAnalyzer.UnitTest/Common/AnalyzerConfigurationTest.cs#L148-L155

File: analyzers/src/SonarAnalyzer.Common/Extensions/AnalyzerOptionsExtensions.cs [hunk 3]
```diff
@@ -33,6 +33,26 @@ public static class AnalyzerOptionsExtensions
...
+                    && path[separatorPosition] is var separator
+                    && separator != Path.DirectorySeparatorChar
+                    && separator != Path.AltDirectorySeparatorChar)
```
Comment (martin-strecker-sonarsource) [inline analyzers/src/SonarAnalyzer.Common/Extensions/AnalyzerOptionsExtensions.cs:49]: We should not take any assumptions about the path separator because `AdditionalFiles` is just a csc.exe command line parameter that was calculated by MSBuild.

**Review by cristian-ambrosini-sonarsource — APPROVED** [review]
LGTM

File: analyzers/src/SonarAnalyzer.Common/Extensions/AnalyzerOptionsExtensions.cs [hunk 4]
```diff
@@ -33,6 +33,26 @@ public static class AnalyzerOptionsExtensions
...
+                {
+                    continue;
+                }
+                return additionalText;
```
Comment (pavel-mikula-sonarsource) [inline analyzers/src/SonarAnalyzer.Common/Extensions/AnalyzerOptionsExtensions.cs:53]: I would flip the condition to remove all the negations and `return addtionalText` inside the `if`. It will make it easier to follow.

File: analyzers/src/SonarAnalyzer.Common/Extensions/AnalyzerOptionsExtensions.cs [hunk 5]
```diff
@@ -33,6 +33,26 @@ public static class AnalyzerOptionsExtensions
...
+    {
+        // HotPath: This code path needs to be allocation free. Don't use Linq.
+        foreach (var additionalText in options.AdditionalFiles) // Uses the struct enumerator of ImmutableArray
+        {
+            // Don't use Path.GetFilename. It allocates a string.
+            if (additionalText.Path is { } path
+                && path.EndsWith(fileName, StringComparison.OrdinalIgnoreCase))
+            {
+                // The character before the filename (if there is a character) must be a directory separator
+                if (path.Length - fileName.Length - 1 is >= 0 and var separatorPosition
```
Comment (pavel-mikula-sonarsource) [inline analyzers/src/SonarAnalyzer.Common/Extensions/AnalyzerOptionsExtensions.cs:46]: This line is abusing the pattern matching, making hit hard to follow. Extract the `var separatorIndex` before the `if` and then just use it the normal way.

File: analyzers/src/SonarAnalyzer.Common/Extensions/AnalyzerOptionsExtensions.cs [hunk 6]
```diff
@@ -33,6 +33,26 @@ public static class AnalyzerOptionsExtensions
...
+                    && path[separatorPosition] is var separator
```
Comment (pavel-mikula-sonarsource) [inline analyzers/src/SonarAnalyzer.Common/Extensions/AnalyzerOptionsExtensions.cs:47]: This `var separator` isn't really needed. The `path[separatorIndex]` doesn't cost anything so I'd use that twice.

**Review by pavel-mikula-sonarsource — APPROVED** [review]
LGTM conceptually. I've left some readability comments

File: analyzers/src/SonarAnalyzer.Common/Extensions/AnalyzerOptionsExtensions.cs [hunk 7]
```diff
@@ -33,6 +33,26 @@ public static class AnalyzerOptionsExtensions
     public static AdditionalText ProjectOutFolderPath(this AnalyzerOptions options) =>
         options.AdditionalFile("ProjectOutFolderPath.txt");
 
-    private static AdditionalText AdditionalFile(this AnalyzerOptions options, string fileName) =>
-        options.AdditionalFiles.FirstOrDefault(x => x.Path is not null && Path.GetFileName(x.Path).Equals(fileName, StringComparison.OrdinalIgnoreCase));
+    private static AdditionalText AdditionalFile(this AnalyzerOptions options, string fileName)
+    {
+        // HotPath: This code path needs to be allocation free. Don't use Linq.
+        foreach (var additionalText in options.AdditionalFiles) // Uses the struct enumerator of ImmutableArray
```
Comment (sharwell) [inline analyzers/src/SonarAnalyzer.Common/Extensions/AnalyzerOptionsExtensions.cs:39]: 💡 We would typically document this in Roslyn with the following attribute on the method:

```
[PerformanceSensitive(
    "https://github.com/SonarSource/sonar-dotnet/issues/7440",
    AllowCaptures = false,
    AllowGenericEnumeration = false)]
```

Comment (sharwell) [reply analyzers/src/SonarAnalyzer.Common/Extensions/AnalyzerOptionsExtensions.cs:39]: It should be possible to use this attribute if you install the [Microsoft.CodeAnalysis.PerformanceSensitiveAnalyzers](https://www.nuget.org/packages/Microsoft.CodeAnalysis.PerformanceSensitiveAnalyzers) package

Comment (sharwell) [reply analyzers/src/SonarAnalyzer.Common/Extensions/AnalyzerOptionsExtensions.cs:39]: Currently we don't have analyzer enforcement for the constraints of the attribute, but we hope to provide it in the future

Comment (martin-strecker-sonarsource) [reply analyzers/src/SonarAnalyzer.Common/Extensions/AnalyzerOptionsExtensions.cs:39]: Thank you @sharwell for the hint.
I added #7690 to our backlog. We will discuss this internally.
